src/iterator.c

```diff
@@ -1181,27 +1189,46 @@ GIT_INLINE(bool) filesystem_iterator_examine_path(
 {
 	bool is_dir = 0;
 	iterator_pathlist_search_t match = ITERATOR_PATHLIST_FULL;
+	int error;
 
-	*is_dir_out = false;
+	*expected_dir = false;
 	*match_out = ITERATOR_PATHLIST_NONE;
 
 	if (iter->base.start_len) {
 		int cmp = iter->base.strncomp(path, iter->base.start, path_len);
 
+		if (cmp < 0)
+			return false;
+
 		/* we haven't stat'ed `path` yet, so we don't yet know if it's a
 		 * directory or not.  special case if the current path may be a
 		 * directory that matches the start prefix.
 		 */
-		if (cmp == 0) {
-			if (iter->base.start[path_len] == '/')
+		if (cmp == 0 && iter->base.start[path_len] == '/') {
+			*expected_dir = true;
+		} else if (cmp == 0) {
+			/* Suppose iter->base.start is "b.c" and path is "b", which is a directory.
```

Comment:
Took me quite some time to track down this bug.

The implementation of diffing in gitstatusd sidesteps this whole business with special handling of slashes. Instead of organizing index as a list of file paths, like it's done in libgit2:

```
  foo/bar/baz.c
  foo/bar/baz.h
```

gitstatusd splits index into directories:

```
foo/bar:
  baz.c
  baz.h
```

To diff a directory in the index against a directory in workdir, we need to `open()` the directory, `fstatat()` every file in it that is listed in the index (this will give us all modified, deleted and type-changed deltas), list all entries under the directory and compare them with the list of entries in the index (this will give us untracked files). Note that nowhere we need to distinguish between file entries and subdirectory entries. We also never need to compare file paths with slashes -- we only compare basenames. This results in simpler and much faster code. Diffing no longer requires comparing strings with long prefixes like it's currently done in `git_diff__from_iterators`.